### PR TITLE
Beamsearch 3.0

### PR DESCRIPTION
**Pull request by Kelina**
Beam search added to the seq2seq decoder.
Also, the decoder got different train/valid/test modes (see comment in the class).

### CODE REVIEW

File: jiant/models.py [hunk 1]
```diff
@@ -1003,15 +1006,24 @@ def _seq_gen_forward(self, batch, task, predict):
...
             target = batch["targs"]["words"][:, 1:].contiguous()
             target_mask = out["target_mask"]
-            logits = out["logits"]
-            task.update_metrics(logits, target, target_mask[:, 1:].contiguous())
 
+            assert "logits" in out or "predictions" in out
+
+            if "logits" in out:
+                logits = out["logits"]
+                task.update_metrics(logits, target, target_mask[:, 1:].contiguous())
```
Comment (W4ngatang) [inline jiant/models.py]: slight preference for using keyword args to make sure the order of args is right, e.g. `(logits=logits, mask=target_mask...)` and `(target=target, ...)`

File: jiant/modules/beam_search.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    Parameters
+    ----------
+    end_index : ``int``
+        The index of the "stop" or "end" token in the target vocabulary.
+    max_steps : ``int``, optional (default = 50)
+        The maximum number of decoding steps to take, i.e. the maximum length
+        of the predicted sequences.
+    beam_size : ``int``, optional (default = 10)
+        The width of the beam used.
+    per_node_beam_size : ``int``, optional (default = beam_size)
```
Comment (W4ngatang) [inline jiant/modules/beam_search.py]: huh, interesting

File: jiant/modules/seq2seq_decoder.py [hunk 3]
```diff
@@ -148,9 +208,16 @@ def forward(
         """
         batch_size, _, _ = encoder_outputs.size()
 
-        use_gold = target_tokens is not None and self.training
+        if self.training:
+            assert target_tokens
+            mode = "train"
+        elif target_tokens is not None:
+            assert target_tokens
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: why would this assert ever fail given the above condition ?

File: jiant/modules/seq2seq_decoder.py [hunk 4]
```diff
@@ -191,16 +273,16 @@ def forward(
...
-            output_dict["loss"] = loss
-        else:
-            # This is needed in case no gold target sequence is available.
-            output_dict["loss"] = torch.tensor([-1.0])
+        targets = target_tokens["words"]
+        target_mask = get_text_field_mask(target_tokens)
+        output_dict["target_mask"] = target_mask
+        relevant_logits = logits[:, : targets.shape[1] - 1, :].contiguous()
+        loss = self._get_loss(relevant_logits, targets, target_mask)
+        output_dict["loss"] = loss
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: 1) Is there a way to return loss of gold tokens at test time ?

2) Do we never query `output_dict` for the loss at test time ? We're no longer hard-setting it to zero here.

File: jiant/tasks/seq2seq.py [hunk 5]
```diff
@@ -118,10 +119,22 @@ def get_metrics(self, reset=False):
...
 
-    def update_metrics(self, logits, labels, tagmask=None):
-        # Cut logits, since generation is done until max seq length (usually >> target length).
-        relevant_logits = logits.max(dim=2)[1][:, : labels.shape[1]]
-        self.scorer2(relevant_logits, labels, tagmask)
+    def update_metrics(self, logits, labels, tagmask=None, predictions=None):
+        assert logits is not None or predictions is not None
+
+        if logits is not None:
+            # Cut logits, since generation is done until max seq length (usually >> target length).
```
Comment (W4ngatang) [inline jiant/tasks/seq2seq.py]: there's no way to stop generation if all sequences in the batch generate EOS ?

File: jiant/tasks/seq2seq.py [hunk 6]
```diff
@@ -118,10 +119,22 @@ def get_metrics(self, reset=False):
...
+            relevant_logits = logits.max(dim=2)[1][:, : labels.shape[1]]
+            self.scorer2(relevant_logits, labels, tagmask)
+        else:
```
Comment (W4ngatang) [inline jiant/tasks/seq2seq.py]: quick comment: `# logits = None, predictions is not None`

File: jiant/trainer.py [hunk 7]
```diff
@@ -851,7 +851,9 @@ def _calculate_validation_performance(
             batch_num += 1
             with torch.no_grad():
                 out = self._forward(batch, task=task)
-            loss = out["loss"]
+            if "loss" in out:
```
Comment (W4ngatang) [inline jiant/trainer.py]: when do we not have a loss ?

Comment (W4ngatang) [reply jiant/trainer.py]: I guess in some seq2seq example, but why can't/wouldn't we compute loss ?

File: jiant/trainer.py [hunk 8]
```diff
@@ -862,16 +864,19 @@ def _calculate_validation_performance(
...
                         gold = batch["targs"]["words"][0][1:]
-                        logits = out["logits"]
-                        output = logits.max(2)[1][0]
-                        input_string, gold_string, output_string = task.get_prediction(
-                            voc_src, voc_trg, inputs, gold, output
-                        )
-                        log.info("\tInput:\t%s", input_string)
-                        log.info("\tGold:\t%s", gold_string)
-                        log.info("\tOutput:\t%s", output_string)
+                        # logits = out["logits"]
```
Comment (W4ngatang) [inline jiant/trainer.py]: delete

also, is `batch_num == 1` just so you don't print every batch ?

File: jiant/modules/seq2seq_decoder.py [hunk 9]
```diff
@@ -148,9 +208,16 @@ def forward(
...
+            assert target_tokens
+            mode = "train"
+        elif target_tokens is not None:
+            assert target_tokens
+            mode = "valid"
+        else:
+            mode = "test"
 
-        if use_gold:
+        if mode == "train" or mode == "valid":
```
Comment (pruksmhc) [inline jiant/modules/seq2seq_decoder.py]: Nit: valid -> validation

File: jiant/tasks/seq2seq.py (hunk 5 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: There is a way: the beam search class currently does that. However, we have no way of knowing beforehand that the generated sequence(s) will not be longer than targets.

File: jiant/modules/seq2seq_decoder.py (hunk 3 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: True, took this out. Thanks!

File: jiant/tasks/seq2seq.py (hunk 6 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: Done.

File: jiant/trainer.py (hunk 8 above)
Comment (Kelina) [reply jiant/trainer.py]: Done. Also: yes, I just wanted one example to see what's going on.

File: jiant/modules/seq2seq_decoder.py (hunk 9 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Done.

File: jiant/models.py (hunk 1 above)
Comment (Kelina) [reply jiant/models.py]: Done.

File: jiant/modules/seq2seq_decoder.py (hunk 4 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Loss is initialized somewhere else in the code to 0.00; this is returned in case we return the beam_search_output earlier.

File: jiant/trainer.py (hunk 7 above)
Comment (Kelina) [reply jiant/trainer.py]: Nono, never when getting validation performance; this is from before when I didn't compute a loss. Changed!

File: jiant/models.py [hunk 10]
```diff
@@ -1003,15 +1006,26 @@ def _seq_gen_forward(self, batch, task, predict):
...
+            assert "logits" in out or "predictions" in out
+
+            if "logits" in out:
+                logits = out["logits"]
+                task.update_metrics(
+                    logits=logits, labels=target, tagmask=target_mask[:, 1:].contiguous()
+                )
+            else:
+                task.update_metrics(
+                    None, target, target_mask[:, 1:].contiguous(), out["predictions"]
```
Comment (W4ngatang) [inline jiant/models.py]: same here with keyword args :) ?

File: jiant/modules/seq2seq_decoder.py [hunk 11]
```diff
@@ -129,6 +138,48 @@ def _initalize_hidden_context_states(self, encoder_outputs, encoder_outputs_mask
...
+            decoder_input, (state["decoder_hidden"], state["decoder_context"])
+        )
+
+        # output projection
+        proj_input = self._projection_bottleneck(state["decoder_hidden"])
+        # (batch_size, num_classes)
+        output_projections = self._output_projection_layer(proj_input)
+
+        # (batch_size, num_classes)
+        step_logit = output_projections
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: what's the point of this line ?

File: jiant/modules/seq2seq_decoder.py [hunk 12]
```diff
@@ -161,15 +227,30 @@ def forward(
...
         step_logits = []
 
         for timestep in range(num_decoding_steps):
-            if use_gold:
-                input_choices = targets[:, timestep]
-            elif timestep == 0:
-                input_choices = torch.ones((batch_size,)) * self._start_index
-            else:
-                input_choices = torch.max(step_logits[-1], dim=2)[1].squeeze(1)
+            input_choices = targets[:, timestep]
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: what happened to the `timestep == 0` case ?

File: jiant/modules/seq2seq_decoder.py [hunk 13]
```diff
@@ -138,7 +189,16 @@ def forward(
...
         # pylint: disable=arguments-differ
         """
-        Decoder logic for producing the entire target sequence at train time.
+        Forward pass for the decoder.
+
+        There are 3 possible modes:
+        1) Training: all metrics are computed given the last gold target tokens at each step
+        2) Validation [no training, but target tokens given]:
+               loss is computed given the gold target tokens;
+               accuracy/BLEU/etc. are computed using predicted tokens, no beam search
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: isn't beam search being used (I guess with `k=1`, but the overall point being that we are also returning model generations in the validation setting) ?

**Comment by W4ngatang** [comment]
Just to check my understanding, the seq2seq module has three modes: train, validation, test.

In train, we only run teacher forcing (use gold tokens as inputs, throw out model outputs)
In validation, we do both teacher forcing and beam search (greedy) to get a loss + sample generation
In test, we only do beam search

Is that right ?

File: jiant/modules/seq2seq_decoder.py [hunk 14]
```diff
@@ -148,9 +208,15 @@ def forward(
...
+            mode = "train"
+        elif target_tokens is not None:
+            mode = "validation"
+        else:
+            mode = "test"
 
-        if use_gold:
+        if mode == "train" or mode == "validation":
             targets = target_tokens["words"]
             target_sequence_length = targets.size()[1]
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: just combine these two lines? `target_sequence_length = targets.size(1) - 1`

File: jiant/modules/seq2seq_decoder.py [hunk 15]
```diff
@@ -148,9 +208,15 @@ def forward(
...
+            assert target_tokens
+            mode = "train"
+        elif target_tokens is not None:
+            mode = "validation"
+        else:
+            mode = "test"
 
-        if use_gold:
+        if mode == "train" or mode == "validation":
             targets = target_tokens["words"]
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: I would prefer to just remove this from the `if` condition, and then do something like `target_sequence_length = targets.size(1) - 1 if mode in ["train", "validation"] else self._max_decoding_steps`.

My mild concern is the presence of lots of code blocks that aren't totally obvious what they each do. This would get rid of a block.

File: jiant/modules/seq2seq_decoder.py (hunk 12 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: timestep == 0 was needed for the case *without* gold targets only. This is handled earlier now.

File: jiant/modules/seq2seq_decoder.py (hunk 13 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Yes, beam search is used during validation; but with beam 1

File: jiant/models.py (hunk 10 above)
Comment (Kelina) [reply jiant/models.py]: Done.

File: jiant/modules/seq2seq_decoder.py (hunk 11 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Absolutely none, I am not sure why it was there. :)

**Comment by Kelina** [comment]
Alex, this is not totally right. (See comments in the forward method of the seq2seq_decoder.py).
However, I changed this slightly. Here both versions for comparison.

*Before your review*:
        There are 3 possible modes:
        1) Training: all metrics are computed given the last gold target tokens at each step
        2) Validation [no training, but target tokens given]:
               loss is computed given the gold target tokens;
               accuracy/BLEU/etc. are computed using predicted tokens, no beam search
        3) Test [no training, no target tokens given]:
               loss is NOT computed (the returned loss is 0.0);
               accuracy/BLEU/etc. are computed using predicted tokens, beam search is used

*Now*:
        1) Training/validation/test with gold data [if generate is False]:
               loss is computed given the gold target tokens;
               accuracy/BLEU/etc. are computed using predicted tokens, no beam search
        2) Generation [if generate is True]:
               loss is NOT computed (the returned loss is 0.0) if no target is given,
                   otherwise as above
               accuracy/BLEU/etc. are computed using predicted tokens, beam search is used

File: jiant/modules/seq2seq_decoder.py (hunk 14 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Done together with your comment below.

File: jiant/modules/seq2seq_decoder.py (hunk 15 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Done.
